[Sema] Preserve ContainsUnexpandedParameterPack in TransformLambdaExpr

zyn0217 opened the pull request:
The lambda `ContainsUnexpandedParameterPack` flag is used for the expressions' dependency computing and is therefore essential for pack expansion. We previously lost the flag's preservation during the lambda's transform, which caused some issues, e.g. a fold expression couldn't properly expand inside a template.

This patch alleviates the issue by retaining the flag in more scenarios. Note that we still have problems with constraints involving packs regarding lambdas, and dealing with that would take more effort, and we'd like to fix them in the future.

Fixes https://github.com/llvm/llvm-project/issues/56852
Fixes https://github.com/llvm/llvm-project/issues/85667
Mitigates https://github.com/llvm/llvm-project/issues/99877 because the attributes were not handled in this patch.

zyn0217 committed:
[Sema] Preserve ContainsUnexpandedParameterPack in TransformLambdaExpr

zyn0217 committed:
Consider the captured pack variables - gh18873

zyn0217 committed:
Merge branch 'main' into GH85667

zyn0217 committed:
comments

zyn0217 committed:
format

zyn0217 committed:
Merge branch 'main' into GH85667

zyn0217 committed:
Retain the 'unexpanded' Decls

zyn0217 committed:
Update ReleaseNotes

zyn0217 committed:
Cleanup & silence the unused-expression warning

zyn0217 committed:
Fix the CI

zyn0217 committed:
Fix typos

zyn0217 committed:
Merge branch 'main' into GH85667

zyn0217 committed:
Fixup

zyn0217 committed:
Fix release notes

zyn0217 committed:
clarify a bit more

cor3ntin commented:
What you have make sense i think. Let me know when you remove the draft status so i can start reviewing the details

zyn0217 commented:
I was waiting for the CI, but it seems Linux CI would take hours. However that shouldn't block the review I think, because Windows CI is green now.

zyn0217 commented on clang/test/SemaTemplate/lambda-capture-pack.cpp:
```diff
@@ -23,3 +23,104 @@ namespace PR41576 {
+    })(1);
+  }(2, 'b');
+
+#if 0
+  // https://github.com/llvm/llvm-project/issues/18873
+  [](auto ...x) { // #1
+    ([&](auto ...y) {  // #2
+      ([x, y] { }(), ...); // #3
+    })(1, 'a');  // #4
+  }(2, 'b');  // #5
+
```
I'm unsure if it's feasible to sneak in these tests, though they still crash for a different reason after this patch. @cor3ntin lmk what you think.

zyn0217 replied:
The FIXME was added to the lines below.

zyn0217 commented on clang/lib/Sema/TreeTransform.h:
```diff
@@ -14636,6 +14645,20 @@ TreeTransform<Derived>::TransformLambdaExpr(LambdaExpr *E) {
+    llvm::SmallVector<UnexpandedParameterPack> UnexpandedPacks;
+    getSema().collectUnexpandedParameterPacksFromLambda(NewCallOperator,
+                                                        UnexpandedPacks);
+    // FIXME: Should we call Sema::DiagnoseUnexpandedParameterPacks() instead?
+    // Unfortunately, that requires the LambdaScopeInfo to exist, which has been
+    // removed by ActOnFinishFunctionBody().
+    LSICopy.ContainsUnexpandedParameterPack = !UnexpandedPacks.empty();
+  }
```
We probably need a refactor here to teach `ActOnFinishFunctionBody()` not to pop the LSI eventually. How about leaving it as an NFC?

mizvekov replied:
I am not sure I understand why we would need to call `Sema::DiagnoseUnexpandedParameterPacks` here.

I think we should be able to diagnose all unexpanded parameter packs during the initial lambda parsing, so what gives?

Also, I am not sure what you mean by leaving it as an NFC.

Regardless, you have a copy of the LSI here, so what would stop you from pushing a new Lambda Scope and copying this LSI back into it? Something like:

```C++
  LambdaScopeInfo *LSI = getSema().PushLambdaScope();
  *LSI = LSICopy;
  Sema::FunctionScopeRAII FuncScopeCleanup(getSema());
```

This is obviously an ugly hack, but I don't think the solution is to simply teach `ActOnFinishFunctionBody` not to pop the function scope.

See this FIXME for a potentially more comprehensive fix: https://github.com/llvm/llvm-project/blob/77ac07444d32668d5826ef27c24180fb10425213/clang/lib/Sema/TreeTransform.h#L14648

zyn0217 replied:
That function attracts me because we seem to 1. always call it after collecting unexpanded packs; 2. also set the `LSI->ContainsUnexpandedParameterPack` flag there. However, this would probably be unnecessary if we called it during parsing.

> Also, I am not sure what you mean by leaving it as an NFC.

I was wondering if we could add an additional flag to `ActOnFinishFunctionBody()` so that we remove the scope info on the caller's own. (e.g. we could do that at the end of `TransformLambdaExpr`.) This way, we could also avoid copying LSI around the lines above. (I had intended to call `ContainsUnexpandedParameterPack()` afterward; again, even if the call is unnecessary, we still could benefit from the refactoring because we don't need to copy LSI.)

> but I don't think the solution is to simply teach ActOnFinishFunctionBody not to pop the function scope.

Can you say more? I think the FIXME is suggesting another way to avoid copying LSI, but that doesn't necessarily mean tweaking `ActOnFinishFunctionBody` is impossible?

zyn0217 replied:
Despite these, I have another concern on performance issue after some deliberation. This is actually a workaround given that we don't have any flags on a `Stmt` to tell whether it contains unexpanded expressions - what we do now leads to visiting the lambda body twice, which might cause some impact in lambda instantiation, if the body is large enough.

A compromise is, we remove this traversal in this patch and *leave it for future fixes*. (Sigh!)

zyn0217 replied:
The double traversal has now been removed from the patch, so I'll mark this resolved.

Endilll commented:
> I was waiting for the CI, but it seems Linux CI would take hours. However that shouldn't block the review I think, because Windows CI is green now.

FWIW Linux CI is recovering as we speak. I expect full recovery over the weekend.

zyn0217 committed:
Typo

ilya-biryukov committed:
[Sema] Default arguments for template parameters affect ContainsUnexpandedPacks

This addresses the FIXME in the code. There are tests for the new
behavior in a follow up fix for #99877, which also addresses other bugs
that prevent exposing the wrong results of `ContainsUnexpandedPacks` in
the outputs of the compiler without crashes.

ilya-biryukov commented:
Just FYI, I wasn't aware of this PR and have ended up with an alternative approach to tracking the unexpanded packs in #99882. (Thanks @cor3ntin for pointing this out)

I'll let the reviewers decide if we should go with this change or mine, also happy to have some combination of the two (e.g. rebase mine on top of this one and still implement some of the refactorings there).

zyn0217 committed:
Merge tests from ilya's 99882 and handle template parameter cases

Co-authored-by: Ilya Biryukov <ibiryukov@google.com>

zyn0217 commented:
I think somehow we still need an extra bit in `Stmt` to carry over the unexpanded info because a body could contain other statements e.g. `DeclStmt` other than full expressions that could propagate up the unexpanded flag through calls to `ContainsUnexpandedParameterPack()`. For example, in one of my test cases:

```cpp
template <class T>
struct identity {
  using type = T;
};


template <class = void> void f() {
  []<class... Is>(Is...) {
    ([] {
      using T = identity<Is>::type;
    }(), ...);
  }(1, 2);
}

template void f(); // #instantiate-f
```

This patch resolves it by visiting the body twice (of course, it's suboptimal), however patch #99882 still crashes because that relies on the mechanism above, I believe.

ilya-biryukov commented:
> This patch resolves it by visiting the body twice (of course, it's suboptimal), however patch #99882 still crashes because that relies on the mechanism above, I believe.

I have also observed it crashes, but I still think that in principle we don't need to traverse the body twice. If `DeclStmt` calls `DiagnoseUnexpandedPacks`, it will set the correct flag on the inner `LambdaScopeInfo`, which should later be available in the `LambdaExpr` itself, and can either be "consumed" by `CXXFoldExpr` (as in the example test case) or propagated into the outer `LamdaScopeInfo` from another call to `DiagnoseUnexpandedPacks`.
I believe that's what's already happening during parsing, but we seem to somehow miss setting this flag during substitution.

Are we crashing because we are missing calls to `DiagnoseUnexpandedPacks` during substitution or because we don't set up the `LambdaScopeInfo` during substitution correctly?

ilya-biryukov commented:
Looking at the code a bit deeper, I think (not 100% sure) we're not calling `DiagnoseUnexpandedPacks` during substitution.
And we probably don't want the logic that produces diagnostics (substitutions cannot create new unexpanded packs in bad places), but we do need the logic to update the `LambdaScopeInfo::ContainsUnexpandedPacks` so we can avoid this extra traversal.

We might want to extract the function that only does the `LambdaScopeInfo::ContainsUnexpandedPacks` and call it instead of the fully-fledged `DiagnoseUnexpandedPacks`, and call it at the right places during substitutions. Tracking down all of them might be tricky, though.

Another alternative is to add a `ContainsUnexpandedPacks` flag to statements, but that is probably even harder than the former suggestion.

The third alternative is to keep the extra traversal for now. It's probably the simplest option.

zyn0217 committed:
Handle more cases e.g. constraints, DeclStmt, etc.

zyn0217 committed:
Merge branch 'main' into GH85667

zyn0217 commented:
> And we probably don't want the logic that produces diagnostics (substitutions cannot create new unexpanded packs in bad places), but we do need the logic to update the LambdaScopeInfo::ContainsUnexpandedPacks so we can avoid this extra traversal.

Cheers. Thanks for the advice. I added logic to TransformDeclStmt so that we can extract such a flag from TypeDecls. (Did I miss anything else? If any, we can probably leave them in the future as problems come up.)

The other challenging part is dealing with constraints. We need somehow to teach TransformCXXFoldExpr to recover the template arguments relative to the primary template so it can properly expand the fold expression. Anyway, I have made it work, though the approach is a bit tricky.

I will add more explanation to the codes tomorrow.

zyn0217 committed:
Co-author

Co-authored-by: Ilya Biryukov <809452+ilya-biryukov@users.noreply.github.com>

zyn0217 commented on clang/lib/Sema/SemaTemplateInstantiate.cpp:
```diff
@@ -1394,7 +1395,22 @@ namespace {
+                                 bool ForConstraints = false) {
+      if (ForConstraints) {
+        LambdaScopeInfo *LSI = getSema().getCurLambda();
+        if (LSI) {
+          MultiLevelTemplateArgumentList MLTAL =
+              getSema().getTemplateInstantiationArgs(
+                  LSI->CallOperator, /*DC=*/nullptr, /*Final=*/false,
+                  /*Innermost=*/std::nullopt, /*RelativeToPrimary=*/true,
+                  /*Pattern=*/nullptr, /*ForConstraintInstantiation=*/true);
+          return getSema().CheckParameterPacksForExpansion(
+              EllipsisLoc, PatternRange, Unexpanded, MLTAL, ShouldExpand,
+              RetainExpansion, NumExpansions);
+        }
```
This is the most convoluted part in this patch. Basically, this targets for a situation where unexpanded packs appear in a trailing constraint. On the basis of our evaluation strategy, we need the complete template arguments, (which are typically relative to the primary template, as opposed to other instantiation algorithm where we usually need the template arguments relative to the parent specializations.) whereas we don't always have them at the time expanding a CXXFoldExpr.

For example, suppose we have
```cpp
template <class = void> void foo() {
 []<class... Is>() {
    ([]()
       requires (!C<Is>)
     {}(),
     ...);
  }.template operator()<char, int, float>();
}
``` 

We would only have one level template arguments, which is `<char, int, float>` when the inner fold expression gets expanded. Since we always preserve the constraint expressions until we evaluate them, we would thereafter have the inner lambda being unexpanded, which is wrong because, later we need to evaluate an expanded constraint when forming CallExprs.

So, one way is to teach the transformation to handle the constraints separately:
If we have only a constraint that contains unexpanded packs, we need to *expand* the expression because we don't actually need to use the constraint immediately. (in other words, the untransformed constraint shouldn't impact the expansion)

And vice versa: if we have combined constraints and other unexpanded packs, we just do the usual: we could hold off on the expansion until we hit the point where only unexpanded constraints are left, which boils down to the above case.

zyn0217 committed:
Fix ReleaseNotes

cor3ntin commented on clang/lib/Sema/TreeTransform.h:
```diff
@@ -4006,6 +4007,37 @@ class TreeTransform {
+    // set as CurContext seems more faithful to the source.
+    TemplateOrNonTemplateCallOperatorDecl->setLexicalDeclContext(Class);
+
+    getDerived().RebuildLambdaExprImpl(StartLoc, EndLoc, LSI);
```
We don't seem to do anything interesting there

zyn0217 replied:
@cor3ntin We call into `TemplateInstantiator::RebuildLambdaExprImpl(),` which substitutes for default arguments. I intend to set the flag after the substitution, as it's likely we could lose the unexpanded pack afterward.

Re "Why not just calling BuildLambdaExpr first?":
`BuildLambdaExpr()` would consume the `ContainsUnexpandedParameterPack` flag and use it as part of the dependency of the `LambdaExpr`. And what we used to do was to substitute for default arguments after `BuildLambdaExpr()`, which makes us miss the chance of altering the flag.

As a result, we would have such a call chain:
```
RebuildLambdaExpr() -> BuildLambdaExprImpl() (which actually performs the default argument substitution) -> BuildLambdaExpr()
```

zyn0217 replied:
@cor3ntin Aha, I've figured out a way to simplify this `RebuildLambdaExpr()` thing: I think it's unnecessary to change the new method decl's lexical context to CurContext (we do so in parsing because we want to align the Scope nesting with the lexical nesting; however this is just what needs to be respected in parsing) and this way, we could just get rid of these setter calls.

cor3ntin commented on clang/lib/Sema/TreeTransform.h:
```diff
@@ -4006,6 +4007,37 @@ class TreeTransform {
+  void RebuildLambdaExprImpl(SourceLocation StartLoc, SourceLocation EndLoc,
+                             LambdaScopeInfo *LSI) {}
+
+  ExprResult RebuildLambdaExpr(SourceLocation StartLoc, SourceLocation EndLoc,
+                               LambdaScopeInfo *LSI) {
+    CXXRecordDecl *Class = LSI->Lambda;
+    CXXMethodDecl *CallOperator = LSI->CallOperator;
+    CallOperator->setLexicalDeclContext(Class);
+    Decl *TemplateOrNonTemplateCallOperatorDecl =
+        CallOperator->getDescribedFunctionTemplate()
+            ? CallOperator->getDescribedFunctionTemplate()
+            : cast<Decl>(CallOperator);
+    // FIXME: Is this really the best choice? Keeping the lexical decl context
+    // set as CurContext seems more faithful to the source.
+    TemplateOrNonTemplateCallOperatorDecl->setLexicalDeclContext(Class);
```
Why not just calling BuildLambdaExpr first?

zyn0217 commented on clang/include/clang/Sema/Sema.h:
```diff
@@ -14037,6 +14037,15 @@ class Sema final : public SemaBase {
       TemplateArgument Arg,
       SmallVectorImpl<UnexpandedParameterPack> &Unexpanded);
 
+  /// Collect the set of unexpanded parameter packs within the given
+  /// template argument.
+  ///
+  /// \param Arg The template argument that will be traversed to find
+  /// unexpanded parameter packs.
+  void collectUnexpandedParameterPacksForFoldExprs(
+      Expr *E, SmallVectorImpl<UnexpandedParameterPack> &Unexpanded,
+      SmallVectorImpl<UnexpandedParameterPack> &UnexpandedFromConstraints);
+
```
I'm using a different API to collect *unexpanded constaints* - they have to be handled individually because we don't transform the constaint expressions like what we have done for other temploids, so typically we don't have sufficient template arguments for them during the transform of CXXFoldExprs. (See my comment https://github.com/llvm/llvm-project/pull/86265#discussion_r1693246465)

cor3ntin replied:
What makes fold expressions different from any other pack expansion?
For example,

```cpp
void f(auto...)
template <class = void> void foo() {
 []<class... Is>() {
    f([]() requires (!C<Is>) {}()...);
  }.template operator()<char, int, float>();
}
```cpp

Isn't that the same scenario?

(and please check the comment, it does look like an incorrect copy paste!)

zyn0217 replied:
> (and please check the comment, it does look like an incorrect copy paste!)
Of course, I'll clean up all the comments (some of them are pretty unclear) until we have an agreement :)

zyn0217 replied:
> For example,
> ...
> Isn't that the same scenario?

(Funny enough, I tried locally and found that we'd never build up a CXXFoldExpr for your example and inevitably... we crash 😅)

Regardless, I'll use my example to demonstrate:

```cpp
template <class = void> void f() {
  []<class... Is>() {
    ([]<class = void>()
       requires (!C<Is>)
     {}(),
     ...);
  }.template operator()<char, int, float>();
}
```

The scenario here is a little subtle. So previously, our expansion for CXXFoldExpr was like:
```cpp
// 1. collect unexpanded parameter packs from the "pattern" so admittedly we would have an unexpanded pack for the constraint
getSema().collectUnexpandedParameterPacks();

// 2. Check whether we should expand with template arguments.
getDerived().TryExpandParameterPacks();

if (!Expand) {
   // 3. Hold off. Rebuild an unexpanded CXXFoldExpr
} else {
   // 4. Otherwise, go on and expand.
}
```

When we want to expand the expression, we're actually in the context of instantiating the body of `foo()`. So we're now given one level of template arguments which is `<void>`.

This is insufficient for the expansion of the inner lambda `[]() requires (!C<Is>) {}`, right? Therefore, we would rebuild the whole pattern, as part of the transform of a callee (`template operator()<char, int, float>()`). Then, we rebuild the call expression and evaluate the constraints there.

However, we would fail to evaluate the constraint because we haven't yet expand the inner pattern, because we don't have sufficient template arguments.

So I take an approach to figure out if we have only unexpanded constraint expressions left. If that is the case, we don't use the current template arguments; instead, we collect them like what we do for constraints: so here we want
two levels of template arguments: `<char, int, float>` and `<void>`.

Using a dedicated version of `collectUnexpandedParameterPacks` could make such this possible as we don't have a way to tell where the unexpanded pack comes from otherwise.

cor3ntin replied:
Let me rephrase. What is the difference between a fold expression and the expansion of parameter pack in a function call or other sort of expressions that could try to expand an outer pack?
Doing something specific for fold expression sounds very wrong

So, I have a concrete suggestion here... can we deal with constraints in a separate PR?
It might help us make progress better.
WDYT?

zyn0217 replied:
> So, I have a concrete suggestion here... can we deal with constraints in a separate PR?
> It might help us make progress better.
> WDYT?

Great. I find I'm getting stuck with it now. Some other cases show up as I'm trying to figure out another approach, so I probably need extra effort here.
Moving forward without the constraints sounds good to me.

zyn0217 committed:
Use CurContext

zyn0217 commented on clang/lib/Sema/SemaTemplateInstantiate.cpp:
```diff
@@ -1673,14 +1690,12 @@ namespace {
           ExprResult ErrorResult = SemaRef.CreateRecoveryExpr(
-              UninstExpr->getBeginLoc(), UninstExpr->getEndLoc(),
-              { UninstExpr }, UninstExpr->getType());
+              UninstExpr->getBeginLoc(), UninstExpr->getEndLoc(), {UninstExpr},
+              UninstExpr->getType());
```
Yeah, an overkill of clang-format..

zyn0217 commented on clang/lib/Sema/SemaTemplateVariadic.cpp:
```diff
@@ -136,6 +155,11 @@ namespace {
     /// do not contain unexpanded parameter packs.
     bool TraverseStmt(Stmt *S) {
       Expr *E = dyn_cast_or_null<Expr>(S);
+
+      llvm::SaveAndRestore _(InConstraint);
+      if (CurrentFunction && CurrentFunction->getTrailingRequiresClause() == S)
+        InConstraint = true;
+
```
Not yet. I'll try to make up one (what a dreadful situation if it gets combined with fold expressions!)

zyn0217 committed:
Remove the fixes for constraints

zyn0217 committed:
Remove extra blanks

zyn0217 committed:
Remove changes for constraints in TransformLambdaExpr

zyn0217 committed:
Simplify the RebuildLambdaExpr()

zyn0217 commented on clang/lib/Sema/SemaLambda.cpp:
```diff
@@ -1105,6 +1107,8 @@ void Sema::ActOnLambdaExpressionAfterIntroducer(LambdaIntroducer &Intro,
 
   CXXMethodDecl *Method = CreateLambdaCallOperator(Intro.Range, Class);
   LSI->CallOperator = Method;
+  // Temporarily set the lexical declaration context to the current
+  // context, so that the Scope stack matches the lexical nesting.
   Method->setLexicalDeclContext(CurContext);
```
I recovered this comment from https://reviews.llvm.org/D124351 because at least it tries to explain something - better than nothing that just makes the lambda parsing dimmer.

cor3ntin commented:
I might be stupid but i still do not get why fold expressions are special, they should not be.
Can you add tests for 

```cpp
void f(auto...);
template <class = void> void foo() {
 []<class... Is>() {
    f([]() requires (!C<Is>) {}()...);
  }.template operator()<char, int, float>();
}
```

other interesting expansions might be 

```cpp
f<[]() requires (!C<Is>) {}()...>();
//
f({+[]() requires (!C<Is>) {}()});
//
[ ...fs = +[]() requires (!C<Is>) {}()]() {
  (fs(),...);
}
}

zyn0217 commented:
> I might be stupid but i still do not get why fold expressions are special, they should not be.

Sorry I was inaccurate suggesting fold expressions are special, it's actually *constraints involving pack expansions* who are special. (I admit inventing another `collectUnexpandedParameterPacks` overload for `CXXFoldExpr` is odd, probably insufficent for the solution because we seem to have more problems - as I observed a similar issue from your first example as well - regarding pack expansion in terms of the constraints on lambdas.)

> 
> ```cpp
> void f(auto...);
> template <class = void> void foo() {
> []<class... Is>() {
>    f([]() requires (!C<Is>) {}()...);
>  }.template operator()<char, int, float>();
>}
> ```

This is awkward: no compiler currently compiles it. (https://gcc.godbolt.org/z/rce3rh7jK) And spuriously, both gcc and clang think that there are no unexpanded parameter packs for expansion. 

Regardless of whether this is conforming, the diagnostic here is simple to explain: we didn't preserve the `ContainsUnexpandedParameterPack` flag while transforming the LambdaExpr, so as an intuitive solution, we could just retain the flag from `TransformLambdaExpr`:

(Though I didn't add the following to the patch yet, it doesn't solve the problem and instead would cause a crash)

```cpp
Expr *TrailingRequires = E->getCallOperator()->getTrailingRequiresClause();
  if (TrailingRequires)
    LSI->ContainsUnexpandedParameterPack |=
        TrailingRequires->containsUnexpandedParameterPack();
```

We still fail to evaluate the constraint because it ends up being dependent even with proper template arguments. This is because the expression is not expanded anywhere before evaluating the constraint, which is one step when we forming up a CallExpr.

The situation could be slightly different (or rather, better?) if a fold expression gets involved. We would still have a chance to expand the fold expression into an expanded but unsubstituted state. So we can make them into the constraint evaluation and substitute arguments there, and get a correct result. So this is why it's convoluted: we probably miss out on some passes for constraint packs: Maybe we need to expand them somewhere before evaluation, but I'm not completely clear.

> ```cpp
> f<[]() requires (!C<Is>) {}()...>();
> //
> f({+[]() requires (!C<Is>) {}()...});
> // I am truly sorry.
> [ ...fs = +[]() requires (!C<Is>) {}()]() {
>  (fs(),...);
> }
> ```

(I promise I won't forget them but let's consider them when we get around to the constraints in the next patch. :)

cor3ntin commented:
Okay so we seem to agree. Can you go over the pr and make sure "fold expression" says "pack expansion" everywhere that it should (release note, comments, function names) ? Thanks!

zyn0217 committed:
Merge branch 'main' into GH85667

zyn0217 committed:
Remove most self-explanatory comments

zyn0217 committed:
Remove one more gratuitous comment

cor3ntin commented on clang/test/SemaCXX/fold_lambda_with_variadics.cpp:
```diff
@@ -0,0 +1,181 @@
+  }(2, 'b');
+
+#if 0
+  // https://github.com/llvm/llvm-project/issues/18873
```
```suggestion
  // FIXME: https://github.com/llvm/llvm-project/issues/18873
```

cor3ntin reviewed (approved):
LGTM (modulo nit)
We should probably backport that

zyn0217 committed:
FIXME

Co-authored-by: cor3ntin <corentinjabot@gmail.com>

zyn0217 committed:
Fix CI

zyn0217 commented:
Thanks for the review! I've just merged it.

> We should probably backport that

SGTM but given that this bug persists in so many versions, I think it is also fine to leave the fix just on trunk - especially since this is actually a partial fix.

We could issue a backport to 19 after we have addressed remaining FIXMEs i.e. bugs on attributes/constraints/captures, if I can catch up the release schedule. WDYT?

cor3ntin commented:
r3 is in 2 weeks and we have to be more conservatives in our backports, so it really depends on how long it took you and @ilya-biryukov to address these issues and the complexity of the patches.

In particular, I would be surprised if the attribute case impacts many people.

ilya-biryukov commented:
> r3 is in 2 weeks and we have to be more conservatives in our backports, so it really depends on how long it took you and @ilya-biryukov to address these issues and the complexity of the patches.
> 
> In particular, I would be surprised if the attribute case impacts many people.

To be clear, I don't have a real-world use-case for attributes. The only reason I have added it in my patch is because there was a test in Clang that started failing with my approach (because it started looking at thing structurally even on the parsing path, as discussed earlier).
I suspect we could live with that for a long time, although fixing that eventually is a good idea. (And is still in my TODO list, just slipped a little)
